src/scanner.c

```diff
@@ -0,0 +1,593 @@
+        return scan_encapsed_part_string(scanner, lexer,
+                                         /* is_after_variable */ false,
+                                         /* is_heredoc */ false,
+                                         /* is_execution_string */ true);
+    }
+
+    if (valid_symbols[ENCAPSED_STRING_CHARS_AFTER_VARIABLE_HEREDOC]) {
+        lexer->result_symbol = ENCAPSED_STRING_CHARS_AFTER_VARIABLE_HEREDOC;
+        return scan_encapsed_part_string(scanner, lexer,
+                                         /* is_after_variable */ true,
+                                         /* is_heredoc */ true,
+                                         /* is_execution_string */ false);
+    }
+
+    if (valid_symbols[ENCAPSED_STRING_CHARS_HEREDOC]) {
+        lexer->result_symbol = ENCAPSED_STRING_CHARS_HEREDOC;
+        return scan_encapsed_part_string(scanner, lexer,
+                                         /* is_after_variable */ false,
+                                         /* is_heredoc */ true,
+                                         /* is_execution_string */ false);
+    }
+
+    if (valid_symbols[NOWDOC_STRING]) {
+        lexer->result_symbol = NOWDOC_STRING;
+        return scan_nowdoc_string(scanner, lexer);
+    }
+
+    if (valid_symbols[HEREDOC_END]) {
+        lexer->result_symbol = HEREDOC_END;
+        if (scanner->open_heredocs->len == 0)
+            return false;
+
+        Heredoc heredoc = VEC_BACK(scanner->open_heredocs);
+
+        while (iswspace(lexer->lookahead)) {
+            skip(lexer);
+        }
+
+        String *word = scan_heredoc_word(lexer);
+        if (strcmp(word->data, heredoc.word->data) != 0) {
+            STRING_FREE(word);
+            return false;
+        }
+        STRING_FREE(word);
+
+        lexer->mark_end(lexer);
+        VEC_POP(scanner->open_heredocs);
+        return true;
+    }
+
+    if (!scan_whitespace(lexer))
+        return false;
+
+    if (valid_symbols[EOF_TOKEN] && lexer->eof(lexer)) {
+        lexer->result_symbol = EOF_TOKEN;
+        return true;
+    }
+
+    if (valid_symbols[HEREDOC_START]) {
+        lexer->result_symbol = HEREDOC_START;
+        Heredoc heredoc;
+
+        while (iswspace(lexer->lookahead)) {
+            skip(lexer);
+        }
+
+        heredoc.word = scan_heredoc_word(lexer);
+        if (heredoc.word->len == 0) {
+            STRING_FREE(heredoc.word);
+            return false;
+        }
+        lexer->mark_end(lexer);
+
+        VEC_PUSH(scanner->open_heredocs, heredoc);
+        return true;
+    }
+
+    if (valid_symbols[AUTOMATIC_SEMICOLON]) {
+        lexer->result_symbol = AUTOMATIC_SEMICOLON;
+
+        if (lexer->lookahead != '?')
+            return false;
+
+        advance(lexer);
+
+        return lexer->lookahead == '>';
+    }
+
+    return false;
+}
+
+void *tree_sitter_php_external_scanner_create() {
+    Scanner *scanner = calloc(1, sizeof(Scanner));
+    scanner->open_heredocs = calloc(1, sizeof(Heredoc));
```

Comment:
Isn't `open_heredocs` a vector? Why `sizeof(Heredoc)` not `sizeof(heredoc_vec)`?

Also, could we just store the heredoc vec inline, inside of the scanner, instead of allocating it separately?

Reply:
Yes..oops. And yes